### PR TITLE
TMP: Copy of #5718 with just the code changes

### PR DESCRIPTION
**Pull request by mdboom**
Making this PR just to make it easier to review.  Any actual changes will still be made to #5718.

### CODE REVIEW

**Comment by mdboom** [comment]
It's fun to see how much actual source code (not just image data) got blitzed by this PR.

File: examples/pylab_examples/demo_annotation_box.py [hunk 1]
```diff
@@ -65,6 +66,7 @@
     arr_lena = read_png(fn)
 
     imagebox = OffsetImage(arr_lena, zoom=0.2)
+    imagebox.image.axes = ax
```
Comment (jenshnielsen) [inline examples/pylab_examples/demo_annotation_box.py]: Since you are touching this file could you change the `arr_lena` name to something else? It really does not make any sense anymore.

File: lib/matplotlib/backends/backend_agg.py [hunk 2]
```diff
@@ -321,7 +321,7 @@ def option_scale_image(self):
         """
         agg backend support arbitrary scaling of image.
         """
```
Comment (jenshnielsen) [inline lib/matplotlib/backends/backend_agg.py]: Should the docstring be updated with this change

File: examples/pylab_examples/demo_annotation_box.py (hunk 1 above)
Comment (mdboom) [inline examples/pylab_examples/demo_annotation_box.py]: Good point.

File: lib/matplotlib/backends/backend_agg.py (hunk 2 above)
Comment (mdboom) [inline lib/matplotlib/backends/backend_agg.py]: Yes.

File: lib/matplotlib/backends/backend_pdf.py [hunk 3]
```diff
@@ -1257,23 +1257,22 @@ def _unpack(self, im):
...
-            alpha = None
-        else:
-            alpha = np.array(alpha, order='C')
-        if im.is_grayscale:
-            r, g, b = rgb.astype(np.float32).transpose(2, 0, 1)
-            gray = (0.3 * r + 0.59 * g + 0.11 * b).astype(np.uint8)[..., None]
-            return h, w, gray, alpha
+        h, w = im.shape[:2]
+        im = im[::-1]
+        if len(im.shape) == 2:
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_pdf.py]: `im.ndim`?

File: lib/matplotlib/backends/backend_svg.py [hunk 4]
```diff
@@ -813,75 +815,57 @@ def draw_image(self, gc, x, y, im, dx=None, dy=None, transform=None):
...
 
         if transform is None:
+            w = 72.0 * w / self.image_dpi
+            h = 72.0 * h / self.image_dpi
+
             self.writer.element(
                 'image',
-                x=short_float_fmt(x/trans[0]),
-                y=short_float_fmt((self.height-y)/trans[3]-h),
+                transform = generate_transform([
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py]: There shouldn't be any spaces around the `=` since this is a keyword argument.

File: lib/matplotlib/backends/backend_wx.py [hunk 5]
```diff
@@ -280,10 +280,8 @@ def draw_image(self, gc, x, y, im):
             b = 0
             w = self.width
             h = self.height
-        rows, cols, image_str = im.as_rgba_str()
-        image_array = np.fromstring(image_str, np.uint8)
-        image_array.shape = rows, cols, 4
-        bitmap = wxc.BitmapFromBuffer(cols, rows, image_array)
+        rows, cols = im.shape[:2]
+        bitmap = wxc.BitmapFromBuffer(cols, rows, im.tobytes())
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_wx.py]: Shouldn't this be `im.tostring()`? `im.tobytes()` is NumPy 1.9.

File: lib/matplotlib/image.py [hunk 6]
```diff
@@ -56,11 +56,131 @@
...
+        renderer, parent, dsu, suppress_composite=None):
+    """
+    Draw a sorted list of artists, compositing images into a single
+    image where possible.
+
+    For internal matplotlib use only: It is here to reduce duplication
+    between `Figure.draw` and `Axes.draw`, but otherwise should not be
+    generally useful.
+    """
+    has_images = np.any(isinstance(x[1], _ImageBase) for x in dsu)
```
Comment (QuLogic) [inline lib/matplotlib/image.py]: I don't believe `np.any` is necessary here.

File: lib/matplotlib/image.py [hunk 7]
```diff
@@ -56,11 +56,131 @@
...
+
+    # override the renderer default if self.suppressComposite
```
Comment (QuLogic) [inline lib/matplotlib/image.py]: There's no `self` here.

File: lib/matplotlib/image.py [hunk 8]
```diff
@@ -162,203 +274,168 @@ def changed(self):
...
-
-    def _get_unsampled_image(self, A, image_extents, viewlim):
+    def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
+                    unsampled=False, round_to_pixel_border=True):
         """
-        convert numpy array A with given extents ([x1, x2, y1, y2] in
-        data coordinate) into the Image, given the viewlim (should be a
-        bbox instance).  Image will be clipped if the extents is
-        significantly larger than the viewlim.
+        Normalize, rescale and color the image `A` to the given
```
Comment (QuLogic) [inline lib/matplotlib/image.py]: The "to" should be "from" here, I think?

File: lib/matplotlib/tests/test_agg.py [hunk 9]
```diff
@@ -155,6 +159,140 @@ def test_long_path():
     fig.savefig(buff, format='png')
 
 
+@image_comparison(baseline_images=['agg_filter'],
+                  extensions=['png'], remove_text=True)
+def test_agg_filter():
+    def smooth1d(x, window_len):
+        # copied from http://www.scipy.org/Cookbook/SignalSmooth
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_agg.py]: This link is no longer correct. It just goes to the cookbook homepage.

File: lib/matplotlib/tests/test_agg.py [hunk 10]
```diff
@@ -155,6 +159,140 @@ def test_long_path():
...
+            if color is None:
+                self.color = (0, 0, 0)
+            else:
+                self.color = color
+
+        def get_pad(self, dpi):
+            return int(self.sigma*3/72.*dpi)
+
+        def process_image(self, padded_src, dpi):
+            #offsetx, offsety = int(self.offsets[0]), int(self.offsets[1])
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_agg.py]: Space after `#` (or just delete.)

File: lib/matplotlib/tests/test_image.py [hunk 11]
```diff
@@ -151,16 +160,26 @@ def test_imsave_color_alpha():
...
-    # Recreate that here:
-    for j in range(3):
-        data[data[:, :, 3] == 0, j] = 1
+    # Recreate the float -> uint8 conversion of the data
+    # We can only expect to be the same with 8 bits of precision,
+    # since that's what the PNG file used.
+    data = (255*data).astype('uint8')
+    arr_buf = (255*arr_buf).astype('uint8')
 
     assert_array_equal(data, arr_buf)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_image.py]: Another blank line needed here.

File: lib/matplotlib/tests/test_image.py [hunk 12]
```diff
@@ -151,16 +160,26 @@ def test_imsave_color_alpha():
...
+    plt.figure()
+
+    np.random.seed(0)
+    Z = np.random.rand(6, 6)
+
+    plt.subplot(121)
+    plt.imshow(Z, alpha=0.5, interpolation='none')
+
+    plt.subplot(122)
+    plt.imshow(Z, alpha=0.5, interpolation='nearest')
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_image.py]: And one here.

File: lib/matplotlib/tests/test_image.py [hunk 13]
```diff
@@ -526,6 +550,85 @@ def test_load_from_url():
...
+                  remove_text=True)
+def test_log_scale_image():
+    Z = np.zeros((10, 10))
+    Z[::2] = 1
+
+    fig = plt.figure()
+    ax = fig.add_subplot(111)
+
+    ax.imshow(Z, extent=[1, 100, 1, 100], cmap='viridis',
+              vmax=abs(Z).max(), vmin=-abs(Z).max())
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_image.py]: Don't you already know what the maximum is?

File: lib/matplotlib/backends/backend_pdf.py (hunk 3 above)
Comment (mdboom) [inline lib/matplotlib/backends/backend_pdf.py]: Done.

File: lib/matplotlib/backends/backend_svg.py (hunk 4 above)
Comment (mdboom) [inline lib/matplotlib/backends/backend_svg.py]: Done.

File: lib/matplotlib/backends/backend_wx.py (hunk 5 above)
Comment (mdboom) [inline lib/matplotlib/backends/backend_wx.py]: Good point.

File: lib/matplotlib/image.py (hunk 6 above)
Comment (mdboom) [inline lib/matplotlib/image.py]: You mean just built-in `any` instead?  That's probably right.

File: lib/matplotlib/tests/test_agg.py (hunk 9 above)
Comment (mdboom) [inline lib/matplotlib/tests/test_agg.py]: I'll just remove the link...

File: lib/matplotlib/image.py (hunk 7 above)
Comment (mdboom) [inline lib/matplotlib/image.py]: Fixed.

File: lib/matplotlib/image.py (hunk 8 above)
Comment (mdboom) [inline lib/matplotlib/image.py]: Yes.

File: lib/matplotlib/tests/test_agg.py (hunk 10 above)
Comment (mdboom) [inline lib/matplotlib/tests/test_agg.py]: Deleted...

File: lib/matplotlib/tests/test_image.py (hunk 11 above)
Comment (mdboom) [inline lib/matplotlib/tests/test_image.py]: This is consistent with the rest of the file.  Better leave purely PEP8 updates to a separate PR.

File: lib/matplotlib/tests/test_image.py (hunk 12 above)
Comment (mdboom) [inline lib/matplotlib/tests/test_image.py]: Ditto.

File: lib/matplotlib/tests/test_image.py (hunk 13 above)
Comment (mdboom) [inline lib/matplotlib/tests/test_image.py]: Good point.

File: lib/matplotlib/image.py (hunk 6 above)
Comment (tacaswell) [inline lib/matplotlib/image.py]: I am some what surprised that `np.any` worked with a generator at all.

Comment (WeatherGod) [inline lib/matplotlib/image.py]: It doesn't:
```
>>> import numpy as np
>>> np.any(False for _ in range(4))
<generator object <genexpr> at 0x7f098d735a00>
```

Comment (mdboom) [inline lib/matplotlib/image.py]: Thanks, guys.  Good catch.  I think we get something that evaluates always as True...

File: lib/matplotlib/tests/test_image.py (hunk 11 above)
Comment (QuLogic) [inline lib/matplotlib/tests/test_image.py]: Oh, I though I saw some other (probably PEP8-related) newlines added here already.

**Comment by QuLogic** [comment]
#5718 was merged.
